ractor_safe_table.c

```diff
@@ -0,0 +1,323 @@
+    if (RUBY_ATOMIC_VALUE_LOAD(*table_obj_ptr) != old_table_obj) {
+        return;
+    }
+
+    struct ractor_safe_table *old_table = RTYPEDDATA_GET_DATA(old_table_obj);
+
+    // This may overcount by up to the number of threads concurrently attempting to insert
+    // GC may also happen between now and the table being rebuilt
+    int expected_size = RUBY_ATOMIC_LOAD(old_table->size) - old_table->deleted_entries;
+
+    struct ractor_safe_table_entry *old_entries = old_table->entries;
+    int old_capacity = old_table->capacity;
+    int new_capacity = old_capacity * 2;
+    if (new_capacity > expected_size * 8) {
+        new_capacity = old_capacity / 2;
+    }
+    else if (new_capacity > expected_size * 4) {
+        new_capacity = old_capacity;
+    }
+
+    // May cause GC and therefore deletes, so must hapen first.
+    VALUE new_table_obj = rb_ractor_safe_table_new(old_table->funcs, new_capacity);
+    struct ractor_safe_table *new_table = RTYPEDDATA_GET_DATA(new_table_obj);
+
+    for (int i = 0; i < old_capacity; i++) {
+        struct ractor_safe_table_entry *entry = &old_entries[i];
+        VALUE key = RUBY_ATOMIC_VALUE_EXCHANGE(entry->key, RACTOR_SAFE_TABLE_MOVED);
+        RUBY_ASSERT(key != RACTOR_SAFE_TABLE_MOVED);
+
+        if (key < RACTOR_SAFE_TABLE_SPECIAL_VALUE_COUNT) continue;
+        if (rb_objspace_garbage_object_p(key)) continue;
+
+        VALUE hash = RUBY_ATOMIC_VALUE_LOAD(entry->hash);
+        if (hash == 0) {
+            // Either in-progress insert or extremely unlikely 0 hash.
+            // Re-calculate the hash.
+            hash = old_table->funcs->hash(key);
+        }
+        RUBY_ASSERT(hash == old_table->funcs->hash(key));
+
+        // Insert key into new_table.
+        struct ractor_safe_table_probe probe;
+        int idx = ractor_safe_table_probe_start(&probe, new_table, hash);
+
+        while (true) {
+            struct ractor_safe_table_entry *entry = &new_table->entries[idx];
+
+            if (entry->key == RACTOR_SAFE_TABLE_EMPTY) {
+                new_table->size++;
+
+                RUBY_ASSERT(new_table->size < new_table->capacity / 2);
+                RUBY_ASSERT(entry->hash == 0);
+
+                entry->key = key;
+                entry->hash = hash;
+                break;
+            }
+            else {
+                RUBY_ASSERT(entry->key >= RACTOR_SAFE_TABLE_SPECIAL_VALUE_COUNT);
+            }
+
+            idx = ractor_safe_table_probe_next(&probe);
+        }
+    }
+
+    RUBY_ATOMIC_VALUE_SET(*table_obj_ptr, new_table_obj);
+
+    RB_GC_GUARD(old_table_obj);
+}
+
+static void
+ractor_safe_table_try_resize(VALUE old_table_obj, VALUE *table_obj_ptr)
+{
+    RB_VM_LOCKING() {
+        ractor_safe_table_try_resize_without_locking(old_table_obj, table_obj_ptr);
+    }
+}
+
+VALUE
+rb_ractor_safe_table_find_or_insert(VALUE *table_obj_ptr, VALUE key, void *data)
+{
+    bool inserting = false;
+    VALUE table_obj;
+
```

Comment:
Just for future proofing new uses of this table, let's add

```suggestion
RUBY_ASSERT(key >= RACTOR_SAFE_TABLE_SPECIAL_VALUE_COUNT);
```